Data obtained from `https://data.boston.gov/dataset/buildbps-facilities-and-educational-data-for-boston-public-schools`

In [1]:
import pandas as pd
import numpy as np

In [2]:
bos = pd.read_csv('../datasets/buildbps.csv')
bos

SMMA_Identifier SMMA_Only_For_Map                       BPS_School_Name  \
0               031               NaN              Adams, Samuel Elementary   
1               078               NaN    Alighieri, Dante Montessori School   
2              045A               NaN            Another Course to College*   
3               012               NaN  Baldwin Early Learning Pilot Academy   
4               087               NaN             Bates, Phineas Elementary   
..              ...               ...                                   ...   
136            011B               NaN                 West Roxbury Academy*   
137            047B               NaN      West Zone Early Learning Center*   
138             100               NaN          Winship, F. Lyman Elementary   
139             073               NaN             Winthrop, John Elementary   
140             044               NaN    Young Achievers Science & Math K-8   

            BPS_Historical_Name SMMA_Abbreviated_Name  \
0                         Adams                 Adams   
1                     Alighieri             Alighieri   
2                          Taft                  ACC*   
3                   Baldwin ELC               Baldwin   
4                         Bates                 Bates   
..                          ...                   ...   
136    West Roxbury High School         West Roxbury*   
137  Hennigan Elementary School            West Zone*   
138                     Winship               Winship   
139         Winthrop Elementary              Winthrop   
140                   Lewenberg       Young Achievers   

                               BPS_Address BRA_Neighborhood  SMMA_latitude  \
0     165 Webster St East Boston, MA 02128      East Boston      42.365553   
1     37 Gove Street East Boston, MA 02128      East Boston      42.371565   
2      20 Warren Street Brighton, MA 02135          Allston      42.350354   
3          121 Corey Rd Brighton, MA 02135         Brighton      42.342037   
4        426 Beech St Roslindale, MA 02131       Roslindale      42.277663   
..                                     ...              ...            ...   
136  1205 VFW Pkwy, West Roxbury, MA 02132     West Roxbury      42.282206   
137   200 Heath St Jamaica Plain, MA 02130    Jamaica Plain      42.326026   
138       54 Dighton St Brighton, MA 02135         Brighton      42.347723   
139   35 Brookford St Dorchester, MA 02125          Roxbury      42.318387   
140        20 Outlook Rd Mattapan, MA02126         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology  ... SMMA_EA_K8_Adequacy_Cafeteria  \
0        -71.034917  Elementary School  ...                           NaN   
1        -71.037608  Elementary School  ...                           NaN   
2        -71.145582        High School  ...                           NaN   
3        -71.140529     Early Learning  ...                           NaN   
4        -71.135353  Elementary School  ...                           NaN   
..              ...                ...  ...                           ...   
136      -71.174549        High School  ...                           NaN   
137      -71.106821     Early Learning  ...                           NaN   
138      -71.154965  Elementary School  ...                           NaN   
139      -71.075341  Elementary School  ...                           NaN   
140      -71.093909                K-8  ...                          Poor   

    SMMA_EA_K8_Adequacy_Stage SMMA_EA_K8_Adequacy_Medical  \
0                         NaN                         NaN   
1                         NaN                         NaN   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   
..                        ...                         ...   
136                       NaN                         NaN   
137                 

### Columns

One of the greatest challenges in this notebook is going to be examining the columns, filling `NaaNs`, and choosing which columns should be dropped from this dataset. After all, there are **251** columns worth of information for just 141 schools.

One cannot view all the columns in this dataset simply by running the command `bos.columns` -- there are simply too many to display. Let's look at all the columns in batches of 50.

In [3]:
bos.columns[:50]

Index(['SMMA_Identifier', 'SMMA_Only_For_Map', 'BPS_School_Name',
       'BPS_Historical_Name', 'SMMA_Abbreviated_Name', 'BPS_Address',
       'BRA_Neighborhood', 'SMMA_latitude', 'SMMA_longitude', 'SMMA_Typology',
       'SMMA_Educational_Program', 'BPS_Grades_Offered', 'BPS_Property_Status',
       'BPS_Total_GSF', 'BPS_GSF', 'SMMA_Site_SF', 'SMMA_Site_Acres',
       'BPS_Year_Founded ', 'BPS_Year_Built', 'MSBA_Year_Reno', 'SMMA_Era',
       'BPS_Plans', 'BPS_Principal', 'BPS_Open', 'BPS_Close', 'DOE_ID',
       'DOE_Total', 'DOE_PK', 'DOE_K', 'DOE_1', 'DOE_2', 'DOE_3', 'DOE_4',
       'DOE_5', 'DOE_6', 'DOE_7', 'DOE_8', 'DOE_9', 'DOE_10', 'DOE_11',
       'DOE_12', 'DOE_SP', 'GSF / Student Value', 'BPS_KWH',
       ' BPS_Electric_Bill ', 'BPS_Therms', ' BPS_Gas_Bill ',
       'BPS_Water_ft_cubic', ' BPS_Water_Bill ', ' BPS_Total_Energy_Cost '],
      dtype='object')

In [4]:
bos.columns[50:100]

Index(['Climate_Priority', 'Climate_Risk', 'SMMA_FA_Visit_Date',
       'SMMA_FA_Status', 'SMMA_FA_School_Enrollment',
       'SMMA_FA_Enrollment_Building', 'SMMA_FA_Major_Investments',
       'SMMA_FA_Existing_PV', 'SMMA_FA_space_Solar', 'SMMA_FA_Roof_Membrane',
       'SMMA_FA_Facade', 'SMMA_FA_Windows', 'SMMA_FA_Boilers',
       'SMMA_FA_Heating_Distribution_Systems',
       'SMMA_FA_Ventilation_Distribution_Systems',
       'SMMA_FA_Electrical_Service', 'SMMA_FA_Means_Egress',
       'SMMA_FA_FP_Sprinklers', 'SMMA_FA_Life_Safety',
       'SMMA_FA_Security_Entry', 'SMMA_FA_Lighting_Quantity_Control',
       'SMMA_FA_Toilets_Fixtures', 'SMMA_FA_Plumbing_Distribution_Systems',
       'SMMA_FA_Accessibility', 'SMMA_FA_Deterioration_Roof',
       'SMMA_FA_Deterioration_Floor', 'SMMA_FA_Deterioration_Walls_Columns',
       'SMMA_FA_Deterioration_Foundations', 'SMMA_FA_Deterioration_Facade',
       'SMMA_FA_lateral_system_identifiable',
       'SMMA_FA_Overall_Building_Condition', 'SMMA_F

In [5]:
bos.columns[100:150]

Index(['SMMA_FA_Walls_Slopes', 'SMMA_FA_Overall_Site_Condition',
       'SMMA_FA_MHC_Archaeological_Assets', 'SMMA_FA_Mass_Transit',
       'SMMA_FA_Bikable', 'SMMA_FA_Walkable', 'SMMA_FA_Overall_Community_Site',
       'SMMA_FA_Site_suitability_school', 'Visit_Date_Date', 'SMMA_ES_Status',
       'SMMA_EA_Building_originally_designed',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_1',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_3',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
       'SMMA_EA_grade_configuration_best_suited_4_to_6',
       'SMMA_EA_grade_configuration_best_suited_6_to_8',
       'SMMA_EA_grade_configuration_best_suited_7_to_8',
       'SMMA_EA_grade_configuration_best_suited_6_to_12',
       'SMMA_EA_grade_configuration_best_suited_7_to_12',
       'SMMA_EA_grade_configuration_best_suited_9_to_12',
       'SMMA_EA_Building_Ventilation', 'SMMA_EA_Building_Natural_Daylighting

In [6]:
bos.columns[150:200]

Index(['SMMA_EA_ES_Adequacy_Music', 'SMMA_EA_ES_Adequacy_Gymnasium',
       'SMMA_EA_ES_Adequacy_Media_Center', 'SMMA_EA_ES_Adequacy_Cafeteria',
       'SMMA_EA_ES_Adequacy_Stage', 'SMMA_EA_ES_Adequacy_Medical',
       'SMMA_EA_ES_Adequacy_Administration', 'SMMA_EA_ES_Adequacy_Custodial',
       'SMMA_EA_ES_Adequacy_Network', 'SMMA_EA_ES_Adequacy_Other_1',
       'SMMA_EA_ES_Adequacy_Other_2', 'SMMA_EA_ES_Adequacy_Other_3',
       'SMMA_EA_ES_Overall_EFE_spaces', 'SMMA_EA_MS_Adequacy_Classroom',
       'SMMA_EA_MS_Adequacy_Science', 'SMMA_EA_MS_Adequacy_SPED_SS',
       'SMMA_EA_MS_Adequacy_SPED_Resource', 'SMMA_EA_MS_Adequacy_Art',
       'SMMA_EA_MS_Adequacy_Music', 'SMMA_EA_MS_Adequacy_Vocations',
       'SMMA_EA_MS_Adequacy_Gymnasium', 'SMMA_EA_MS_Adequacy_Media_Center',
       'SMMA_EA_MS_Adequacy_Cafeteria', 'SMMA_EA_MS_Adequacy_Stage',
       'SMMA_EA_MS_Adequacy_Medical', 'SMMA_EA_MS_Adequacy_Administration',
       'SMMA_EA_MS_Adequacy_Custodial', 'SMMA_EA_MS_Adequacy_Network'

In [7]:
bos.columns[200:252]

Index(['SMMA_EA_HS712_Adequacy_Network', 'SMMA_EA_HS712_Adequacy_Other_1',
       'SMMA_EA_HS712_Adequacy_Other_2', 'SMMA_EA_HS712_Adequacy_Other_3',
       'SMMA_EA_HS712_Overall_EFE_spaces', 'SMMA_EA_HS912_Adequacy_Classroom',
       'SMMA_EA_HS912_Adequacy_TeacherPlanning',
       'SMMA_EA_HS912_Adequacy_Small_Group', 'SMMA_EA_HS912_Adequacy_Science',
       'SMMA_EA_HS912_Adequacy_SPED_SS',
       'SMMA_EA_HS912_Adequacy_SPED_Resource', 'SMMA_EA_HS912_Adequacy_Art',
       'SMMA_EA_HS912_Adequacy_Music', 'SMMA_EA_HS912_Adequacy_Vocations',
       'SMMA_EA_HS912_Adequacy_Gymnasium',
       'SMMA_EA_HS912_Adequacy_GynasiumAlternatives',
       'SMMA_EA_HS912_Adequacy_Media_Center',
       'SMMA_EA_HS912_Adequacy_Cafeteria', 'SMMA_EA_HS912_Adequacy_Auditorium',
       'SMMA_EA_HS912_Adequacy_Stage', 'SMMA_EA_HS912_Adequacy_Medical',
       'SMMA_EA_HS912_Adequacy_Administration',
       'SMMA_EA_HS912_Adequacy_Custodial', 'SMMA_EA_HS912_Adequacy_Network',
       'SMMA_EA_HS912_Adequac

#### Where do we begin?

One good place to be would be with the `NaaNs`. If there are any columns with a large about of data missing (i.e., 50% of more) -- dropping these columns from the dataset could be a good place to start.

In [8]:
# Let's look at 100 columns with the most NaaNs
bos.isna().sum().sort_values(ascending = False)[:100]

SMMA_EA_HS712_Adequacy_Network                    141
SMMA_EA_grade_configuration_best_suited_4_to_6    141
SMMA_EA_MS_Adequacy_Network                       141
SMMA_EA_HS712_Adequacy_Other_3                    141
SMMA_EA_K8_Adequacy_Network                       140
                                                 ... 
SMMA_EA_K8_Adequacy_Media_Center                  105
SMMA_EA_K8_Adequacy_Cafeteria                     105
SMMA_EA_K8_Adequacy_Stage                         105
SMMA_EA_K8_Adequacy_SPED_SS                       105
SMMA_EA_K8_Adequacy_Medical                       105
Length: 100, dtype: int64

100 columns in the dataset have **75% or more** missing data! The data in these columns may have been valuable, but I wouldn't feel comfortable filling `NaaNs` with any values (i.e., mean value of the column) with so much data missing. Let's drop these columns from the dataset.

In [9]:
[column for column in bos.columns if bos[column].isna().sum() >= 105]

['SMMA_Only_For_Map',
 'MSBA_Year_Reno',
 'SMMA_EA_grade_configuration_best_suited_4_to_6',
 'SMMA_EA_grade_configuration_best_suited_6_to_8',
 'SMMA_EA_grade_configuration_best_suited_7_to_8',
 'SMMA_EA_grade_configuration_best_suited_6_to_12',
 'SMMA_EA_grade_configuration_best_suited_7_to_12',
 'SMMA_EA_grade_configuration_best_suited_9_to_12',
 'SMMA_EA_ES_Adequacy_Custodial',
 'SMMA_EA_ES_Adequacy_Network',
 'SMMA_EA_ES_Adequacy_Other_1',
 'SMMA_EA_ES_Adequacy_Other_2',
 'SMMA_EA_ES_Adequacy_Other_3',
 'SMMA_EA_MS_Adequacy_Classroom',
 'SMMA_EA_MS_Adequacy_Science',
 'SMMA_EA_MS_Adequacy_SPED_SS',
 'SMMA_EA_MS_Adequacy_SPED_Resource',
 'SMMA_EA_MS_Adequacy_Art',
 'SMMA_EA_MS_Adequacy_Music',
 'SMMA_EA_MS_Adequacy_Vocations',
 'SMMA_EA_MS_Adequacy_Gymnasium',
 'SMMA_EA_MS_Adequacy_Media_Center',
 'SMMA_EA_MS_Adequacy_Cafeteria',
 'SMMA_EA_MS_Adequacy_Stage',
 'SMMA_EA_MS_Adequacy_Medical',
 'SMMA_EA_MS_Adequacy_Administration',
 'SMMA_EA_MS_Adequacy_Custodial',
 'SMMA_EA_MS_Adequac

In [10]:
# Drop columns with over 105 NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= 105])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_K8_Overall_EFE_spaces                        104
SMMA_EA_grade_configuration_best_suited_PreK_to_6    104
Climate_Risk                                          99
Climate_Priority                                      97
SMMA_EA_ES_Adequacy_Art                               92
                                                    ... 
SMMA_FA_Enrollment_Building                            0
SMMA_FA_School_Enrollment                              0
SMMA_FA_Status                                         0
 BPS_Total_Energy_Cost                                 0
SMMA_Identifier                                        0
Length: 151, dtype: int64

Even after dropping 100 columns from our dataset that were riddled with `NaaNs`, we still have 151 columns, some of which still have over 50% `NaaNs` in them (see above). Let's take a look at ant columns with over `50%` of their values missing from the dataset.

In [11]:
[column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2]

['Climate_Priority',
 'Climate_Risk',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
 'SMMA_EA_ES_Adequacy_PreK',
 'SMMA_EA_ES_Adequacy_Kindergarten',
 'SMMA_EA_ES_Adequacy_Classroom',
 'SMMA_EA_ES_Adequacy_SPED_SS',
 'SMMA_EA_ES_Adequacy_SPED_Resource',
 'SMMA_EA_ES_Adequacy_Art',
 'SMMA_EA_ES_Adequacy_Music',
 'SMMA_EA_ES_Adequacy_Gymnasium',
 'SMMA_EA_ES_Adequacy_Media_Center',
 'SMMA_EA_ES_Adequacy_Cafeteria',
 'SMMA_EA_ES_Adequacy_Stage',
 'SMMA_EA_ES_Adequacy_Medical',
 'SMMA_EA_ES_Adequacy_Administration',
 'SMMA_EA_ES_Overall_EFE_spaces',
 'SMMA_EA_K8_Overall_EFE_spaces']

In [12]:
# Drop columns with over 50% NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_grade_configuration_best_suited_PreK_to_3    64
SMMA_EA_grade_configuration_best_suited_PreK_to_1    56
SMMA_EA_Building_originally_designed                 18
DOE_SP                                               17
BPS_Year_Founded                                     17
                                                     ..
SMMA_FA_Overall_Building_Condition                    0
SMMA_FA_Overall_Site_Condition                        0
SMMA_FA_Overall_Community_Site                        0
SMMA_EA_Overall_EFE_learning_environments             0
SMMA_Identifier                                       0
Length: 132, dtype: int64

Ok...last command in which `NaaNs`, I promise.

After looking at the columns that remain in our dataset, we can see that there are still two columns with `64` and `56` `NaaN` values. After that, the next highest `NaaN` count is `18`, which is much more managable.

Let's drop those other 2 columns from our dataset.

In [13]:
# Drop remaining two columns with over 50 NaaNs
bos.drop(columns = ['SMMA_EA_grade_configuration_best_suited_PreK_to_3',
                    'SMMA_EA_grade_configuration_best_suited_PreK_to_1'],
         inplace = True)

In [14]:
# How many columns do we have now?
bos.shape

(141, 130)

### Now that we're done removing columns from the dataset, let's remove some columns from the dataset

In [15]:
drop_columns = []

It's become clear that there are unecessary spaces in multiple `column` names for this dataset. Instead of changing them all one by one, let's run the command below to get rid of them all in one fell swoop.

In [16]:
for column in bos.columns:
    bos.rename(columns = {column : column.replace(' ', '')}, inplace = True)

There are still 130 columns in this dataset, which is a lot to include for just 141 schools. If we drop any more columns from this dataset, we'll have to be more objective.

In [17]:
bos[bos.columns[:20]]

SMMA_Identifier                       BPS_School_Name  \
0               031              Adams, Samuel Elementary   
1               078    Alighieri, Dante Montessori School   
2              045A            Another Course to College*   
3               012  Baldwin Early Learning Pilot Academy   
4               087             Bates, Phineas Elementary   
..              ...                                   ...   
136            011B                 West Roxbury Academy*   
137            047B      West Zone Early Learning Center*   
138             100          Winship, F. Lyman Elementary   
139             073             Winthrop, John Elementary   
140             044    Young Achievers Science & Math K-8   

            BPS_Historical_Name SMMA_Abbreviated_Name  \
0                         Adams                 Adams   
1                     Alighieri             Alighieri   
2                          Taft                  ACC*   
3                   Baldwin ELC               Baldwin   
4                         Bates                 Bates   
..                          ...                   ...   
136    West Roxbury High School         West Roxbury*   
137  Hennigan Elementary School            West Zone*   
138                     Winship               Winship   
139         Winthrop Elementary              Winthrop   
140                   Lewenberg       Young Achievers   

                               BPS_Address BRA_Neighborhood  SMMA_latitude  \
0     165 Webster St East Boston, MA 02128      East Boston      42.365553   
1     37 Gove Street East Boston, MA 02128      East Boston      42.371565   
2      20 Warren Street Brighton, MA 02135          Allston      42.350354   
3          121 Corey Rd Brighton, MA 02135         Brighton      42.342037   
4        426 Beech St Roslindale, MA 02131       Roslindale      42.277663   
..                                     ...              ...            ...   
136  1205 VFW Pkwy, West Roxbury, MA 02132     West Roxbury      42.282206   
137   200 Heath St Jamaica Plain, MA 02130    Jamaica Plain      42.326026   
138       54 Dighton St Brighton, MA 02135         Brighton      42.347723   
139   35 Brookford St Dorchester, MA 02125          Roxbury      42.318387   
140        20 Outlook Rd Mattapan, MA02126         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology SMMA_Educational_Program  \
0        -71.034917  Elementary School   Elementary: Pre-K to 6   
1        -71.037608  Elementary School   Elementary: Pre-K to 6   
2        -71.145582        High School    High Schools: 9 to 12   
3        -71.140529     Early Learning   Elementary: Pre-K to 6   
4        -71.135353  Elementary School   Elementary: Pre-K to 6   
..              ...                ...                      ...   
136      -71.174549        High School    High Schools: 9 to 12   
137      -71.106821     Early Learning   Elementary: Pre-K to 6   
138      -71.154965  Elementary School   Elementary: Pre-K to 6   
139      -71.075341  Elementary School   Elementary: Pre-K to 6   
140      -71.093909                K-8   Elementary: Pre-K to 8   

    BPS_Grades_Offered BPS_Property_Status  BPS_Total_GSF   BPS_GSF  \
0                 K0-5               Owned        41336.0   41336.0   
1                 K0-5               Owned        16948.0   16948.0   
2                 9-12               Owned        28021.0   28021.0   
3                 K0-1               Owned        27879.0   27879.0   
4                 K1-5               Owned        27936.0   27936.0   
..                 ...                 ...            ...       ...   
136               9-12               Owned       113508.0  113508.0   
137               K0-1               Owned        27000.0   27000.0   
138               K0-5               Owned        41016.0   41016.0   
139               K1-5               Owned        40530.0   40530.0   
140               K1-8               Owned       120944.0  1209

In [18]:
one_twenty_drop_columns = ['SMMA_Identifier', 'BPS_Historical_Name', 'SMMA_Abbreviated_Name', 
                           'BPS_Total_GSF', 'SMMA_Site_SF', 'SMMA_Educational_Program','SMMA_Era',
                           'BPS_Property_Status', 'BPS_Total_GSF', 'SMMA_Site_SF', 'BPS_Plans']

for i in one_twenty_drop_columns:
    drop_columns.append(i)
    
drop_columns

['SMMA_Identifier',
 'BPS_Historical_Name',
 'SMMA_Abbreviated_Name',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'SMMA_Educational_Program',
 'SMMA_Era',
 'BPS_Property_Status',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'BPS_Plans']

In [19]:
bos[bos.columns[20:40]]

BPS_Principal    BPS_Open   BPS_Close    DOE_ID  DOE_Total  DOE_PK  \
0          Hannah Irvin  7:30:00 AM  1:30:00 PM  350302.0        302    43.0   
1          Mayra Cuevas  7:30:00 AM  1:30:00 PM  350066.0         90    29.0   
2    Lisa Gilbert-Smith  7:30:00 AM  2:30:00 PM  350541.0        237     0.0   
3            Tavia Mead  7:30:00 AM  4:35:00 PM  350003.0        142    68.0   
4            Kelly Hung  9:30:00 AM  3:30:00 PM  350278.0        297    20.0   
..                  ...         ...         ...       ...        ...     ...   
136                 NaN         NaN         NaN  350574.0        529     0.0   
137       Jean Larrabee  7:30:00 AM  4:35:00 PM  350006.0         93    24.0   
138    Nadine Marcellus  9:30:00 AM  3:30:00 PM  350374.0        258    23.0   
139          Leah Blake  8:30:00 AM  3:00:00 PM  350180.0        349    28.0   
140   Virginia Chalmers  8:30:00 AM  4:00:00 PM  350380.0        542    34.0   

     DOE_K  DOE_1  DOE_2  DOE_3  DOE_4  DOE_5  DOE_6  DOE_7  DOE_8  DOE_9  \
0     49.0   53.0   42.0   48.0   37.0   30.0    0.0    0.0    0.0    0.0   
1     15.0   16.0   12.0    8.0    6.0    4.0    0.0    0.0    0.0    0.0   
2      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   54.0   
3     41.0   33.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4     36.0   39.0   44.0   49.0   58.0   51.0    0.0    0.0    0.0    0.0   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
136    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  136.0   
137   34.0   35.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
138   33.0   42.0   44.0   49.0   34.0   33.0    0.0    0.0    0.0    0.0   
139   60.0   50.0   59.0   65.0   48.0   39.0    0.0    0.0    0.0    0.0   
140   48.0   51.0  115.0   64.0   63.0   34.0   42.0   47.0   44.0    0.0   

     DOE_10  DOE_11  DOE_12  DOE_SP  
0       0.0     0.0     0.0     0.0  
1       0.0     0.0     0.0     0.0  
2      62.0    62.0    59.0     0.0  
3       0.0     0.0     0.0     0.0  
4       0.0     0.0     0.0     0.0  
..      ...     ...     ...     ...  
136   140.0   103.0   147.0     3.0  
137     0.0     0.0     0.0     0.0  
138     0.0     0.0     0.0     0.0  
139     0.0     0.0     0.0     0.0  
140     0.0     0.0     0.0     0.0  

[141 rows x 20 columns]

In [20]:
drop_columns_twenty_forty = ['DOE_ID', 'DOE_Total', 'DOE_1', 'DOE_2', 'DOE_3', 'DOE_4', 'DOE_5', 
                             'DOE_6', 'DOE_7', 'DOE_8', 'DOE_9', 'DOE_10', 'DOE_11', 'DOE_12', 'DOE_SP']

for i in drop_columns_twenty_forty:
    drop_columns.append(i)

drop_columns

['SMMA_Identifier',
 'BPS_Historical_Name',
 'SMMA_Abbreviated_Name',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'SMMA_Educational_Program',
 'SMMA_Era',
 'BPS_Property_Status',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'BPS_Plans',
 'DOE_ID',
 'DOE_Total',
 'DOE_1',
 'DOE_2',
 'DOE_3',
 'DOE_4',
 'DOE_5',
 'DOE_6',
 'DOE_7',
 'DOE_8',
 'DOE_9',
 'DOE_10',
 'DOE_11',
 'DOE_12',
 'DOE_SP']

In [21]:
bos[bos.columns[40:60]]

GSF/StudentValue    BPS_KWH BPS_Electric_Bill  BPS_Therms   BPS_Gas_Bill  \
0               180.0   381000.0       $58,339.00      14650.0    $13,401.00    
1               180.0     9977.0        $2,072.00       8230.0     $9,957.00    
2               225.0   343410.0       $49,657.00      51459.0    $57,978.00    
3               200.0    70212.0        $9,588.00       9806.0    $11,657.00    
4               180.0    91800.0       $13,367.00      14530.0    $16,909.00    
..                ...        ...               ...         ...            ...   
136             225.0  1097520.0      $294,332.00     257019.0   $258,704.00    
137             200.0   120384.0       $74,435.00     267450.0   $276,004.00    
138             180.0   128480.0       $20,522.00      15656.0    $18,100.00    
139             175.0   138240.0       $20,594.00      16897.0    $19,484.00    
140             195.0   385248.0       $53,897.00      36421.0    $56,973.00    

     BPS_Water_ft_cubic BPS_Water_Bill BPS_Total_Energy_Cost  \
0                   0.0          $-              $71,740.00    
1                8120.0     $1,685.00            $13,714.00    
2              163110.0    $20,763.00           $128,398.00    
3               42550.0     $5,227.00            $26,472.00    
4               42870.0     $5,262.00            $35,538.00    
..                  ...            ...                   ...   
136            265450.0    $36,183.00           $589,219.00    
137            287880.0    $37,653.00           $388,092.00    
138             44180.0     $5,394.00            $44,016.00    
139             53440.0     $7,457.00            $47,535.00    
140             90940.0    $11,457.00           $122,327.00    

    SMMA_FA_Visit_Date SMMA_FA_Status  SMMA_FA_School_Enrollment  \
0           05/18/2016        Phase 3                        302   
1           05/18/2016        Phase 3                         90   
2           04/27/2016        Phase 3                        237   
3           04/25/2016        Phase 3                        142   
4           06/08/2016        Phase 3                        297   
..                 ...            ...                        ...   
136         06/01/2016        Phase 3                        529   
137         04/26/2016        Phase 3                         93   
138         05/10/2016        Phase 3                        258   
139         06/10/2016        Phase 3                        349   
140         06/11/2016        Phase 3                        542   

     SMMA_FA_Enrollment_Building SMMA_FA_Major_Investments  \
0                            302                       Yes   
1                             90                       Yes   
2                            674                       Yes   
3                            142                       Yes   
4                            297                       Yes   
..                           ...                       ...   
136                         1026                       Yes   
137                          717                        No   
138                          258                       Yes   
139                          349                       Yes   
140                          542                       Yes   

    SMMA_FA_Existing_PV SMMA_FA_space_Solar SMMA_FA_Roof_Membrane  \
0                    No                 Yes              Moderate   
1                    No                 Yes                 Minor   
2                    No                 Yes                 Minor   
3                    No                 Yes           Not Present   
4                    No                 Yes              Moderate   
..                  ...                 ...                   ...   
136                  No                 Yes               Replace   
137                  No                 Yes              Moderate   
138                  No                 Yes              Moderate   
139        

In [22]:
drop_columns_forty_sixty = ['BPS_Water_ft_cubic', 'SMMA_FA_Visit_Date', 'SMMA_FA_Status', 
                            'SMMA_FA_School_Enrollment', 'SMMA_FA_Enrollment_Building', 
                            'SMMA_FA_Major_Investments', 'SMMA_FA_Existing_PV', 'SMMA_FA_space_Solar', 
                            'SMMA_FA_Roof_Membrane', 'SMMA_FA_Facade', 'SMMA_FA_Windows', 'SMMA_FA_Boilers', 
                            'SMMA_FA_Heating_Distribution_Systems', 'SMMA_FA_Electrical_Service', 
                            'SMMA_FA_Ventilation_Distribution_Systems', 'SMMA_FA_Electrical_Service']

for i in drop_columns_forty_sixty:
    drop_columns.append(i)
    
drop_columns

['SMMA_Identifier',
 'BPS_Historical_Name',
 'SMMA_Abbreviated_Name',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'SMMA_Educational_Program',
 'SMMA_Era',
 'BPS_Property_Status',
 'BPS_Total_GSF',
 'SMMA_Site_SF',
 'BPS_Plans',
 'DOE_ID',
 'DOE_Total',
 'DOE_1',
 'DOE_2',
 'DOE_3',
 'DOE_4',
 'DOE_5',
 'DOE_6',
 'DOE_7',
 'DOE_8',
 'DOE_9',
 'DOE_10',
 'DOE_11',
 'DOE_12',
 'DOE_SP',
 'BPS_Water_ft_cubic',
 'SMMA_FA_Visit_Date',
 'SMMA_FA_Status',
 'SMMA_FA_School_Enrollment',
 'SMMA_FA_Enrollment_Building',
 'SMMA_FA_Major_Investments',
 'SMMA_FA_Existing_PV',
 'SMMA_FA_space_Solar',
 'SMMA_FA_Roof_Membrane',
 'SMMA_FA_Facade',
 'SMMA_FA_Windows',
 'SMMA_FA_Boilers',
 'SMMA_FA_Heating_Distribution_Systems',
 'SMMA_FA_Electrical_Service',
 'SMMA_FA_Ventilation_Distribution_Systems',
 'SMMA_FA_Electrical_Service']

In [23]:
bos[bos.columns[60:80]]

SMMA_FA_Ventilation_Distribution_Systems SMMA_FA_Electrical_Service  \
0                                    Replace                      Minor   
1                                   Adequate                   Adequate   
2                                   Moderate                   Adequate   
3                                      Minor                      Minor   
4                                   Moderate                    Replace   
..                                       ...                        ...   
136                                    Minor                   Adequate   
137                                    Minor                      Minor   
138                                 Moderate                   Moderate   
139                                  Replace                   Moderate   
140                                 Moderate                      Minor   

    SMMA_FA_Means_Egress SMMA_FA_FP_Sprinklers SMMA_FA_Life_Safety  \
0            Not Present           Not Present               Minor   
1               Moderate              Adequate            Adequate   
2                  Minor           Not Present            Adequate   
3                  Minor           Not Present            Adequate   
4               Adequate           Not Present               Minor   
..                   ...                   ...                 ...   
136             Adequate              Moderate               Minor   
137                Minor           Not Present               Minor   
138             Moderate           Not Present            Adequate   
139             Moderate              Adequate               Minor   
140             Adequate           Not Present            Adequate   

    SMMA_FA_Security_Entry SMMA_FA_Lighting_Quantity_Control  \
0                 Moderate                             Minor   
1                    Minor                          Adequate   
2                 Moderate                             Minor   
3                 Moderate                             Minor   
4                 Moderate                          Adequate   
..                     ...                               ...   
136               Adequate                             Minor   
137               Moderate                          Adequate   
138                  Minor                             Minor   
139               Moderate                           Replace   
140               Moderate                             Minor   

    SMMA_FA_Toilets_Fixtures SMMA_FA_Plumbing_Distribution_Systems  \
0                    Replace                               Replace   
1                   Adequate                              Adequate   
2                   Adequate                              Adequate   
3                   Moderate                              Moderate   
4                    Replace                              Moderate   
..                       ...                                   ...   
136                 Moderate                              Moderate   
137                 Moderate                              Moderate   
138                    Minor                              Moderate   
139                  Replace                               Replace   
140                    Minor                              Moderate   

    SMMA_FA_Accessibility SMMA_FA_Deterioration_Roof  \
0             Not Present               Not Observed   
1                 Replace               Not Observed   
2                Moderate               Not Observed   
3                   Minor               Not Observed   
4                 Replace               Not Observed   
..                    ...                        ...   
136              Adequate               Not Observed   
137                 Minor               Not Observed   
138               Replace               Not Observed   
139               Replace               Not Observed   
140               Replace           

In [24]:
drop_columns_sixty_eighty = ['SMMA_FA_Ventilation_Distribution_Systems', 'SMMA_FA_Electrical_Service', 
                             'SMMA_FA_Means_Egress', 'SMMA_FA_FP_Sprinklers', 'SMMA_FA_Life_Safety',
                             'SMMA_FA_Security_Entry', 'SMMA_FA_Lighting_Quantity_Control',
                             'SMMA_FA_Toilets_Fixtures', 'SMMA_FA_Plumbing_Distribution_Systems',
                             'SMMA_FA_Accessibility', 'SMMA_FA_Deterioration_Roof', 'SMMA_FA_Deterioration_Floor',
                             'SMMA_FA_Deterioration_Walls_Columns', 'SMMA_FA_Deterioration_Foundations', 
                             'SMMA_FA_Deterioration_Facade', 'SMMA_FA_lateral_system_identifiable',
                             'SMMA_FA_MHC_Inventory', 'SMMA_FA_MHC_Register', 'SMMA_FA_Emergency_Shelter']

for i in drop_columns_sixty_eighty:
    drop_columns.append(i)

In [25]:
bos[bos.columns[80:100]]

SMMA_FA_Community_Use_spaces SMMA_FA_Overall_Community_Building  \
0                             No                               Poor   
1                            Yes                               Good   
2                             No                               Poor   
3                             No                               Poor   
4                             No                               Poor   
..                           ...                                ...   
136                          Yes                               Good   
137                          Yes                               Good   
138                           No                               Poor   
139                           No                               Poor   
140                          Yes                               Good   

    SMMA_FA_Building_suitability_school  \
0                                   Yes   
1                                   Yes   
2                                   Yes   
3                                   Yes   
4                                   Yes   
..                                  ...   
136                                 Yes   
137                                 Yes   
138                                 Yes   
139                                 Yes   
140                                 Yes   

    SMMA_FA_susceptible_climate_change_now  \
0                                       No   
1                                      Yes   
2                                       No   
3                                       No   
4                                       No   
..                                     ...   
136                                     No   
137                                     No   
138                                     No   
139                                     No   
140                                     No   

    SMMA_FA_susceptible_climate_change_2100  \
0                                        No   
1                                       Yes   
2                                        No   
3                                        No   
4                                        No   
..                                      ...   
136                                      No   
137                                      No   
138                                      No   
139                                      No   
140                                      No   

    SMMA_FA_site_major_investments_10__years  \
0                                         No   
1                                        Yes   
2                                         No   
3                                         No   
4                                         No   
..                                       ...   
136                                      Yes   
137                                       No   
138                                       No   
139                                       No   
140                                       No   

    SMMA_FA_building_expandable_current_site SMMA_FA_Parking_Quality  \
0                                        Yes                Moderate   
1                                         No                Moderate   
2                                         No                Adequate   
3                                         No                   Minor   
4                                        Yes                 Replace   
..                                       ...                     ...   
136                                      Yes                Adequate   
137                                       No                   Minor   
138                                       No                 Replace   
139                                      Yes             Not Present   
140                                      Yes                Moderate   

    SMMA_FA_Neighborhood_Streets SMMA_FA_DropOff_PickUp  \
0              

In [26]:
drop_columns_eighty_hundred = ['SMMA_FA_Community_Use_spaces', 'SMMA_FA_Overall_Community_Building', 
                               'SMMA_FA_Building_suitability_school', 'SMMA_FA_susceptible_climate_change_now',
                               'SMMA_FA_susceptible_climate_change_2100', 
                               'SMMA_FA_site_major_investments_10__years',
                               'SMMA_FA_building_expandable_current_site',  
                               'SMMA_FA_DropOff_PickUp', 'SMMA_FA_Walkways_Curbs_Sidewalks', 
                               'SMMA_FA_MAABADA_Accessibility', 'SMMA_FA_Site_Lighting', 'SMMA_FA_Fencing', 
                               'SMMA_FA_Drainage', 'SMMA_FA_Play_Areas', 'SMMA_FA_Walls_Slopes', 
                               'SMMA_FA_MHC_Archaeological_Assets']

for i in drop_columns_eighty_hundred:
    drop_columns.append(i)

In [27]:
bos[bos.columns[100:120]]

SMMA_FA_Bikable SMMA_FA_Walkable SMMA_FA_Overall_Community_Site  \
0               Yes              Yes                           Fair   
1               Yes              Yes                           Fair   
2               Yes              Yes                           Good   
3               Yes              Yes                           Good   
4               Yes              Yes                           Fair   
..              ...              ...                            ...   
136              No               No                           Good   
137             Yes              Yes                           Good   
138             Yes              Yes                           Fair   
139             Yes              Yes                           Fair   
140             Yes              Yes                           Fair   

    SMMA_FA_Site_suitability_school Visit_Date_Date SMMA_ES_Status  \
0                               Yes         5/18/16        Phase 3   
1                                No         5/18/16        Phase 3   
2                               Yes         4/27/16        Phase 3   
3                                No         4/25/16        Phase 3   
4                               Yes          6/8/16        Phase 3   
..                              ...             ...            ...   
136                             Yes         4/26/16        Phase 3   
137                             Yes         5/10/16        Phase 3   
138                              No         6/10/16        Phase 3   
139                              No             NaN        Phase 3   
140                             Yes             NaN        Phase 3   

    SMMA_EA_Building_originally_designed SMMA_EA_Building_Ventilation  \
0                          Middle School                    Deficient   
1                      Elementary School                    Deficient   
2                            High School                         Poor   
3                      Elementary School                         Poor   
4                      Elementary School                         Poor   
..                                   ...                          ...   
136                          High School                         Fair   
137               Early Education Center                         Fair   
138                    Elementary School                         Poor   
139                    Elementary School                         Poor   
140                          High School                         Poor   

    SMMA_EA_Building_Natural_Daylighting  \
0                                   Fair   
1                                   Fair   
2                                   Good   
3                                   Fair   
4                                   Good   
..                                   ...   
136                                 Good   
137                                 Fair   
138                                 Poor   
139                                 Fair   
140                                 Good   

    SMMA_EA_Building_Lighting_Quality_Quantity SMMA_EA_Building_Air_Quality  \
0                                         Fair                    Deficient   
1                                         Good                         Fair   
2                                         Fair                         Fair   
3                                         Poor                         Poor   
4                                    Excellent                         Poor   
..                                         ...                          ...   
136                                       Fair                         Fair   
137                                       Fair                         Fair   
138                                       Fair                         Poor   
139                                       Fair                         Poor   
140                                

In [28]:
drop_columns_hundred_hundredtwenty = ['SMMA_FA_Site_suitability_school', 'Visit_Date_Date', 'SMMA_ES_Status', 
                                      'SMMA_EA_Building_originally_designed',
                                      'SMMA_EA_Building_Lighting_Quality_Quantity', 
                                      'SMMA_EA_Building_Acoustical', 'SMMA_EA_Building_Technology:_Power', 
                                      'SMMA_EA_Building_Technology_Wireless',
                                      'SMMA_EA_Building_Technology_Interactive', 'SMMA_EA_Building_Furniture',
                                      'SMMA_EA_Building_Finishes', 'SMMA_EA_Building_Outdoor_Classrooms',
                                      'SMMA_EA_Building_Environment_inviting_stimulating_comfortable',
                                      'SMMA_EA_Building_Adjacencies_Learning_Environments']

for i in drop_columns_hundred_hundredtwenty:
    drop_columns.append(i)

In [29]:
bos[bos.columns[120:140]]

SMMA_EA_Overall_EFE_learning_environments  \
0                                        Poor   
1                                        Fair   
2                                        Fair   
3                                        Poor   
4                                        Fair   
..                                        ...   
136                                      Fair   
137                                      Fair   
138                                      Poor   
139                                      Fair   
140                                      Fair   

    SMMA_EA_site_includes_Playgrounds SMMA_EA_site_includes_Accessible  \
0                                 Yes                               No   
1                                 Yes                              Yes   
2                                  No                               No   
3                                 Yes                              Yes   
4                                 Yes                               No   
..                                ...                              ...   
136                                No                              Yes   
137                               Yes                              Yes   
138                               Yes                              Yes   
139                               Yes                              Yes   
140                               Yes                              Yes   

    SMMA_EA_site_includes_PlayFields SMMA_EA_change_typology_easily  \
0                                 No                             No   
1                                 No                             No   
2                                 No                             No   
3                                 No                             No   
4                                 No                            Yes   
..                               ...                            ...   
136                              Yes                            Yes   
137                               No                             No   
138                               No                             No   
139                               No                            NaN   
140                              Yes                            Yes   

    SMMA_EA_transformed_educationally SMMA_EA_swing_space  \
0                                  No                 Yes   
1                                 Yes                  No   
2                                  No                  No   
3                                  No                  No   
4                                  No                 Yes   
..                                ...                 ...   
136                               Yes                 Yes   
137                                No                  No   
138                                No                  No   
139                                No                 Yes   
140                               Yes                 Yes   

    SMMA_EA_85%_to_115%_utilization SMMA_EA_Educational_Program  \
0                               Yes      Elementary: Pre-K to 6   
1                               Yes      Elementary: Pre-K to 6   
2                               Yes       High Schools: 9 to 12   
3                               Yes      Elementary: Pre-K to 6   
4                               Yes      Elementary: Pre-K to 6   
..                              ...                         ...   
136                              No       High Schools: 9 to 12   
137                             Yes      Elementary: Pre-K to 6   
138                             Yes      Elementary: Pre-K to 6   
139                             Yes      Elementary: Pre-K to 6   
140                             Yes      Elementary: Pre-K to 8   

    SMMA_EA_EFE_S_Overall_all  
0                        Poor  
1                        Poor  
2                        Poor  
3      

In [30]:
drop_columns_hundredtwenty_hundredforty = ['SMMA_EA_Overall_EFE_learning_environments', 
                                           'SMMA_EA_site_includes_Accessible', 
                                           'SMMA_EA_change_typology_easily', 
                                           'SMMA_EA_transformed_educationally',
                                           'SMMA_EA_swing_space', 'SMMA_EA_85%_to_115%_utilization',
                                           'SMMA_EA_Educational_Program', 'SMMA_EA_EFE_S_Overall_all']

for i in drop_columns_hundredtwenty_hundredforty:
    drop_columns.append(i)

### How were columns chosen to be dropped? 

Now that we have a list of columns to be dropped from the `bos` dataset, we should ask just how these columns were chosen to be dropped. Admittedly, these columns were chosen a bit subjectively. However, there were several factors that led to these selections:

- `Ambiguity`. This was easily the biggest reason for omitting columns from the dataset. there were so many   columns with titles that were heavily abbreviated and seemed to offer no valuable insight. Also, a lot of the data seems to be heavily subjective, and is rated on a scale that was not provided. It's also entirely possible that the ratings were made by just one individual, further skewing the data to the opinions of that individual. 

- `General unhelpfuness`. Some of the columns in this dataset provided information that simply seemed to be neither helpful nor interesting.

In [33]:
# Drop columns in our drop_columns list
bos = bos.drop(columns = drop_columns)

# Check work
bos.shape

KeyError: "['SMMA_Identifier' 'BPS_Historical_Name' 'SMMA_Abbreviated_Name'\n 'BPS_Total_GSF' 'SMMA_Site_SF' 'SMMA_Educational_Program' 'SMMA_Era'\n 'BPS_Property_Status' 'BPS_Total_GSF' 'SMMA_Site_SF' 'BPS_Plans' 'DOE_ID'\n 'DOE_Total' 'DOE_1' 'DOE_2' 'DOE_3' 'DOE_4' 'DOE_5' 'DOE_6' 'DOE_7'\n 'DOE_8' 'DOE_9' 'DOE_10' 'DOE_11' 'DOE_12' 'DOE_SP' 'BPS_Water_ft_cubic'\n 'SMMA_FA_Visit_Date' 'SMMA_FA_Status' 'SMMA_FA_School_Enrollment'\n 'SMMA_FA_Enrollment_Building' 'SMMA_FA_Major_Investments'\n 'SMMA_FA_Existing_PV' 'SMMA_FA_space_Solar' 'SMMA_FA_Roof_Membrane'\n 'SMMA_FA_Facade' 'SMMA_FA_Windows' 'SMMA_FA_Boilers'\n 'SMMA_FA_Heating_Distribution_Systems' 'SMMA_FA_Electrical_Service'\n 'SMMA_FA_Ventilation_Distribution_Systems' 'SMMA_FA_Electrical_Service'\n 'SMMA_FA_Ventilation_Distribution_Systems' 'SMMA_FA_Electrical_Service'\n 'SMMA_FA_Means_Egress' 'SMMA_FA_FP_Sprinklers' 'SMMA_FA_Life_Safety'\n 'SMMA_FA_Security_Entry' 'SMMA_FA_Lighting_Quantity_Control'\n 'SMMA_FA_Toilets_Fixtures' 'SMMA_FA_Plumbing_Distribution_Systems'\n 'SMMA_FA_Accessibility' 'SMMA_FA_Deterioration_Roof'\n 'SMMA_FA_Deterioration_Floor' 'SMMA_FA_Deterioration_Walls_Columns'\n 'SMMA_FA_Deterioration_Foundations' 'SMMA_FA_Deterioration_Facade'\n 'SMMA_FA_lateral_system_identifiable' 'SMMA_FA_MHC_Inventory'\n 'SMMA_FA_MHC_Register' 'SMMA_FA_Emergency_Shelter'\n 'SMMA_FA_Community_Use_spaces' 'SMMA_FA_Overall_Community_Building'\n 'SMMA_FA_Building_suitability_school'\n 'SMMA_FA_susceptible_climate_change_now'\n 'SMMA_FA_susceptible_climate_change_2100'\n 'SMMA_FA_site_major_investments_10__years'\n 'SMMA_FA_building_expandable_current_site' 'SMMA_FA_DropOff_PickUp'\n 'SMMA_FA_Walkways_Curbs_Sidewalks' 'SMMA_FA_MAABADA_Accessibility'\n 'SMMA_FA_Site_Lighting' 'SMMA_FA_Fencing' 'SMMA_FA_Drainage'\n 'SMMA_FA_Play_Areas' 'SMMA_FA_Walls_Slopes'\n 'SMMA_FA_MHC_Archaeological_Assets' 'SMMA_FA_Site_suitability_school'\n 'Visit_Date_Date' 'SMMA_ES_Status' 'SMMA_EA_Building_originally_designed'\n 'SMMA_EA_Building_Lighting_Quality_Quantity'\n 'SMMA_EA_Building_Acoustical' 'SMMA_EA_Building_Technology:_Power'\n 'SMMA_EA_Building_Technology_Wireless'\n 'SMMA_EA_Building_Technology_Interactive' 'SMMA_EA_Building_Furniture'\n 'SMMA_EA_Building_Finishes' 'SMMA_EA_Building_Outdoor_Classrooms'\n 'SMMA_EA_Building_Environment_inviting_stimulating_comfortable'\n 'SMMA_EA_Building_Adjacencies_Learning_Environments'\n 'SMMA_EA_Overall_EFE_learning_environments'\n 'SMMA_EA_site_includes_Accessible' 'SMMA_EA_change_typology_easily'\n 'SMMA_EA_transformed_educationally' 'SMMA_EA_swing_space'\n 'SMMA_EA_85%_to_115%_utilization' 'SMMA_EA_Educational_Program'\n 'SMMA_EA_EFE_S_Overall_all'] not found in axis"

### Currency Formatting

A few columns report data in the form of currencies. These objects should be housed as `floats` but are instead are stored as `objects` with several symbols and punctuation marks embedded in each cell. The columns are fiull of `NaNs`, and in order to fill those empty cells with a useful value, we're going to have to remove to symbols from each cell of each column and convert the `object` to a float.

The below funciton has been created to do just that.

In [ ]:
# Define function to change values from objects to floats
def object_currency_to_float(column):
    
    # Create empty list
    new_column = []
    
    # Loop through each cell in the inputted column
    for cell in bos[column]:
        
        # Check to see if each cell is a string
        if type(cell) == str:
            
            cell = cell.replace('$', '')
    
            cell = cell.replace('-', '0')
            
            cell = cell.replace(',', '')
                
            cell = float(cell)
            
            new_column.append(cell)
            
        else:
            
            new_column.append(cell)
    
    return new_column
    

**BPS_Electric_Bill**

In [ ]:
# Call function and set equal to the Electric Bill column
bos['BPS_Electric_Bill'] = object_currency_to_float('BPS_Electric_Bill')

In [ ]:
# Fill NaaNs in this column with the average value of Electric Bills that we have available 
#\n in the dataset
bos['BPS_Electric_Bill'].fillna(bos['BPS_Electric_Bill'].mean(), inplace = True)

Let's check to see if we'll have to run the `object_currency_to_float` function for other columns (Hint: we will.).

**BPS_Gas_Bill**

In [ ]:
bos['BPS_Gas_Bill']

In [ ]:
bos['BPS_Gas_Bill'] = object_currency_to_float('BPS_Gas_Bill')

In [ ]:
# Fill NaaNs in this column with the average value of Gas Bills that we have available 
#\n in the dataset
bos['BPS_Gas_Bill'].fillna(round(bos['BPS_Gas_Bill'].mean()), inplace = True)

**BPS_Water_Bill**

In [ ]:
# Call function and set equal to the Electric Bill column
bos['BPS_Water_Bill'] = object_currency_to_float('BPS_Water_Bill')

In [ ]:
# Fill NaaNs in this column with the average value of Gas Bills that we have available 
#\n in the dataset
bos['BPS_Water_Bill'].fillna(round(bos['BPS_Water_Bill'].mean()), inplace = True)

**BPS_Total_Energy_Cost**

In [ ]:
# Call function and set equal to the Electric Bill column
bos['BPS_Total_Energy_Cost'] = object_currency_to_float('BPS_Total_Energy_Cost')

In [ ]:
# Fill NaaNs in this column with the average value of Gas Bills that we have available 
#\n in the dataset
bos['BPS_Total_Energy_Cost'].fillna(round(bos['BPS_Total_Energy_Cost'].mean()), inplace = True)